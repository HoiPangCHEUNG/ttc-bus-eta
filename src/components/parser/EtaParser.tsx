import { parseRoute } from "./routeName";

export const etaParser = (json: any) => {
  const result: {
    line: string;
    stopName: string;
    routeName: string;
    etas: { id: number; second: any; busId: any; branch: String }[];
  }[] = [];

  if (Object.keys(json).length === 0) {
    return [];
  }

  if (Array.isArray(json.body.predictions) === false) {
    console.log("single lines stop");
    if (json.body.predictions["@_dirTitleBecauseNoPredictions"] === undefined) {
      if (Array.isArray(json.body.predictions.direction) === false) {
        // multiple lines => multiple directions
        result.push({
          line: json.body.predictions["@_routeTag"],
          stopName: json.body.predictions["@_stopTitle"],
          routeName: parseRoute(json.body.predictions["@_routeTitle"]),
          etas: [],
        });
        json.body.predictions.direction.prediction.map(
          (element: any, index: number) => {
            result[result.length - 1].etas.push({
              id: index,
              second: element["@_seconds"],
              busId: element["@_vehicle"],
              branch: element["@_branch"],
            });
            return null;
          }
        );
      } else {
        // 1 prediction, 2 directions
        // Eg. stops/14761 returns 939A, 939B
        const line = json.body.predictions["@_routeTag"];
        const title = json.body.predictions["@_stopTitle"];
        json.body.predictions.direction.map((element: any, index: number) => {
          // Only lines with etas are listed
          if (element["@_dirTitleBecauseNoPredictions"] === undefined) {
            result.push({
              line,
              stopName: title,
              routeName: "",
              etas: [],
            });
            if (Array.isArray(element.prediction) === false) {
              result[result.length - 1].etas.push({
                id: 0,
                second: element.direction.prediction["@_seconds"],
                busId: element.direction.prediction["@_vehicle"],
                branch: element.direction.prediction["@_branch"],
              });
            } else {
              element.prediction.map((el2: any, index2: number) => {
                result[result.length - 1].etas.push({
                  id: index2,
                  second: el2["@_seconds"],
                  busId: el2["@_vehicle"],
                  branch: el2["@_branch"],
                });
                return null;
              });
            }
            return null;
          }
          return null;
        });
        // if no line have ETA, keep a title
        if (result.length === 0) {
          result.push({
            line: "No ETAs detected.",
            stopName: json.body.predictions[0]["@_stopTitle"],
            routeName: "",
            etas: [],
          });
        }
        return result;
      }
    } else {
      console.log("no ETA at all");
      result.push({
        line: "No ETAs detected.",
        stopName: json.body.predictions["@_stopTitle"],
        routeName: "",
        etas: [],
      });
    }
  } else {
    console.log("multi line stop");
    json.body.predictions.map((element: any, index: number) => {
      // Only lines with etas are listed
      if (element["@_dirTitleBecauseNoPredictions"] === undefined) {
        if (Array.isArray(element.direction)) {
          const title = element["@_stopTitle"];
          const line = element["@_routeTag"];
          element.direction.map((el3: any, index3: number) => {
            result.push({
              line,
              stopName: title,
              routeName: parseRoute(el3["@_title"]),
              etas: [],
            });
            if (Array.isArray(el3.prediction) === false) {
              console.log("single prediction");
              result[result.length - 1].etas.push({
                id: 0,
                second: el3.prediction["@_seconds"],
                busId: el3.prediction["@_vehicle"],
                branch: el3.prediction["@_branch"],
              });
            } else {
              el3.prediction.map((el2: any, index2: number) => {
                result[result.length - 1].etas.push({
                  id: index2,
                  second: el2["@_seconds"],
                  busId: el2["@_vehicle"],
                  branch: el2["@_branch"],
                });
                return null;
              });
            }

            return null;
          });
        } else {
          result.push({
            line: element["@_routeTag"],
            stopName: element["@_stopTitle"],
            routeName: parseRoute(element["@_routeTitle"]),
            etas: [],
          });
          if (Array.isArray(element.direction.prediction) === false) {
            result[result.length - 1].etas.push({
              id: 0,
              second: element.direction.prediction["@_seconds"],
              busId: element.direction.prediction["@_vehicle"],
              branch: element.direction.prediction["@_branch"],
            });
          } else {
            element.direction.prediction.map((el2: any, index2: number) => {
              result[result.length - 1].etas.push({
                id: index2,
                second: el2["@_seconds"],
                busId: el2["@_vehicle"],
                branch: el2["@_branch"],
              });
              return null;
            });
          }
        }
        return null;
      }
      return null;
    });
    // if no line have ETA, keep a title
    if (result.length === 0) {
      console.log("empty db");
      result.push({
        line: "No ETAs detected.",
        stopName: json.body.predictions[0]["@_stopTitle"],
        routeName: "",
        etas: [],
      });
    }
  }
  return result;
};
