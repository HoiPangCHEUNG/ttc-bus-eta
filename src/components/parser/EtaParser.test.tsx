// import renderer from 'react-test-renderer';
import { etaParser } from "./EtaParser";
import { expect, test } from "@jest/globals";
import renderer from "react-test-renderer";

describe("eta parsing", () => {
  const testValues = [
    {
      title: "no ETA",
      input: {
        body: {
          predictions: [
            {
              "@_agencyTitle": "Toronto Transit Commission",
              "@_routeTitle": "39-Finch East",
              "@_routeTag": "39",
              "@_stopTitle": "480 Gordon Baker Rd",
              "@_stopTag": "7900",
              "@_dirTitleBecauseNoPredictions":
                "West - 39 Finch East towards Finch Station",
            },
            {
              "@_agencyTitle": "Toronto Transit Commission",
              "@_routeTitle": "42-Cummer",
              "@_routeTag": "42",
              "@_stopTitle": "480 Gordon Baker Rd",
              "@_stopTag": "7900",
              "@_dirTitleBecauseNoPredictions":
                "West - 42 Cummer towards Finch Station",
            },
          ],
        },
      },
    },
    {
      title: "1 predictions 1 direction",
      input: {
        body: {
          predictions: {
            direction: {
              prediction: [
                {
                  "@_epochTime": "1672201380997",
                  "@_seconds": "96",
                  "@_minutes": "1",
                  "@_isDeparture": "false",
                  "@_affectedByLayover": "true",
                  "@_branch": "39",
                  "@_dirTag": "39_1_39B",
                  "@_vehicle": "3109",
                  "@_block": "39_67_20",
                  "@_tripTag": "45227730",
                },
                {
                  "@_epochTime": "1672202531994",
                  "@_seconds": "1247",
                  "@_minutes": "20",
                  "@_isDeparture": "false",
                  "@_affectedByLayover": "true",
                  "@_branch": "39",
                  "@_dirTag": "39_1_39B",
                  "@_vehicle": "3264",
                  "@_block": "39_69_30",
                  "@_tripTag": "45227729",
                },
                {
                  "@_epochTime": "1672203731994",
                  "@_seconds": "2447",
                  "@_minutes": "40",
                  "@_isDeparture": "false",
                  "@_affectedByLayover": "true",
                  "@_branch": "39",
                  "@_dirTag": "39_1_39B",
                  "@_vehicle": "9234",
                  "@_block": "39_71_132",
                  "@_tripTag": "45227728",
                },
              ],
              "@_title": "West - 39 Finch East towards Finch Station",
            },
            "@_agencyTitle": "Toronto Transit Commission",
            "@_routeTitle": "39-Finch East",
            "@_routeTag": "39",
            "@_stopTitle": "Old Finch Ave At Morningside Ave",
            "@_stopTag": "12925",
          },
          "@_copyright": "All data copyright Toronto Transit Commission 2022.",
        },
      },
    },
    {
      title: "1 predictions 2 directions",
      input: {
        "?xml": {
          "@_version": "1.0",
          "@_encoding": "utf-8",
        },
        body: {
          predictions: {
            direction: [
              {
                prediction: [
                  {
                    "@_epochTime": "1672202160000",
                    "@_seconds": "611",
                    "@_minutes": "10",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39A",
                    "@_dirTag": "39_0_39A",
                    "@_vehicle": "3124",
                    "@_block": "39_66_10",
                    "@_tripTag": "45227587",
                  },
                  {
                    "@_epochTime": "1672203360000",
                    "@_seconds": "1811",
                    "@_minutes": "30",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39A",
                    "@_dirTag": "39_0_39A",
                    "@_vehicle": "3281",
                    "@_block": "39_72_70",
                    "@_tripTag": "45227586",
                  },
                  {
                    "@_epochTime": "1672204560000",
                    "@_seconds": "3011",
                    "@_minutes": "50",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39A",
                    "@_dirTag": "39_0_39A",
                    "@_vehicle": "3234",
                    "@_block": "39_77_142",
                    "@_tripTag": "45227585",
                  },
                ],
                "@_title": "East - 39a Finch East towards Neilson",
              },
              {
                prediction: [
                  {
                    "@_epochTime": "1672201560000",
                    "@_seconds": "11",
                    "@_minutes": "0",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39B",
                    "@_dirTag": "39_0_39B",
                    "@_vehicle": "9227",
                    "@_block": "39_10_100",
                    "@_tripTag": "45227579",
                  },
                  {
                    "@_epochTime": "1672202760000",
                    "@_seconds": "1211",
                    "@_minutes": "20",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39B",
                    "@_dirTag": "39_0_39B",
                    "@_vehicle": "3283",
                    "@_block": "39_12_122",
                    "@_tripTag": "45227578",
                  },
                  {
                    "@_epochTime": "1672203960000",
                    "@_seconds": "2411",
                    "@_minutes": "40",
                    "@_isDeparture": "true",
                    "@_affectedByLayover": "true",
                    "@_branch": "39B",
                    "@_dirTag": "39_0_39B",
                    "@_vehicle": "3109",
                    "@_block": "39_67_20",
                    "@_tripTag": "45227577",
                  },
                ],
                "@_title":
                  "East - 39b Finch East towards Old Finch and Morningview",
              },
            ],
            "@_agencyTitle": "Toronto Transit Commission",
            "@_routeTitle": "39-Finch East",
            "@_routeTag": "39",
            "@_stopTitle": "Finch Station",
            "@_stopTag": "14211",
          },
          "@_copyright": "All data copyright Toronto Transit Commission 2022.",
        },
      },
    },
    {
      title: "2 predictions 2 direction",
      input: {
        body: {
          predictions: [
            {
              "@_agencyTitle": "Toronto Transit Commission",
              "@_routeTitle": "339-Finch East Night Bus",
              "@_routeTag": "339",
              "@_stopTitle": "Finch Ave East At Wilfred Ave",
              "@_stopTag": "8607",
              "@_dirTitleBecauseNoPredictions":
                "East - 339 Finch East towards Markham Rd",
            },
            {
              direction: [
                {
                  prediction: [
                    {
                      "@_epochTime": "1672202317976",
                      "@_seconds": "338",
                      "@_minutes": "5",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39A",
                      "@_dirTag": "39_0_39A",
                      "@_vehicle": "3124",
                      "@_block": "39_66_10",
                      "@_tripTag": "45227587",
                    },
                    {
                      "@_epochTime": "1672203517976",
                      "@_seconds": "1538",
                      "@_minutes": "25",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39A",
                      "@_dirTag": "39_0_39A",
                      "@_vehicle": "3281",
                      "@_block": "39_72_70",
                      "@_tripTag": "45227586",
                    },
                    {
                      "@_epochTime": "1672204718036",
                      "@_seconds": "2738",
                      "@_minutes": "45",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39A",
                      "@_dirTag": "39_0_39A",
                      "@_vehicle": "3234",
                      "@_block": "39_77_142",
                      "@_tripTag": "45227585",
                    },
                  ],
                  "@_title": "East - 39a Finch East towards Neilson",
                },
                {
                  prediction: [
                    {
                      "@_epochTime": "1672202426996",
                      "@_seconds": "447",
                      "@_minutes": "7",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39B",
                      "@_dirTag": "39_0_39B",
                      "@_vehicle": "9227",
                      "@_block": "39_10_100",
                      "@_tripTag": "45227579",
                    },
                    {
                      "@_epochTime": "1672203268756",
                      "@_seconds": "1289",
                      "@_minutes": "21",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39B",
                      "@_dirTag": "39_0_39B",
                      "@_vehicle": "3283",
                      "@_block": "39_12_122",
                      "@_tripTag": "45227578",
                    },
                    {
                      "@_epochTime": "1672204468756",
                      "@_seconds": "2489",
                      "@_minutes": "41",
                      "@_isDeparture": "false",
                      "@_affectedByLayover": "true",
                      "@_branch": "39B",
                      "@_dirTag": "39_0_39B",
                      "@_vehicle": "3109",
                      "@_block": "39_67_20",
                      "@_tripTag": "45227577",
                    },
                  ],
                  "@_title":
                    "East - 39b Finch East towards Old Finch and Morningview",
                },
              ],
              "@_agencyTitle": "Toronto Transit Commission",
              "@_routeTitle": "39-Finch East",
              "@_routeTag": "39",
              "@_stopTitle": "Finch Ave East At Wilfred Ave",
              "@_stopTag": "8607",
            },
          ],
          "@_copyright": "All data copyright Toronto Transit Commission 2022.",
        },
      },
    },
    {
      title: "empty result",
      input: {},
    },
  ];

  test.each(testValues)("$title", ({ input }) => {
    const component = renderer.create(
      <p>{JSON.stringify(etaParser(input), null, 4)}</p>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
