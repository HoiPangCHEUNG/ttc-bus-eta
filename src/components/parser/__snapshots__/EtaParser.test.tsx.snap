// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eta parsing 1 predictions 1 direction 1`] = `
<p>
  [
    {
        "line": "39",
        "stopName": "Old Finch Ave At Morningside Ave",
        "routeName": "Finch East",
        "etas": [
            {
                "id": 0,
                "second": "96",
                "busId": "3109",
                "branch": "39",
                "tripTag": "45227730",
                "epochTime": "1672201380997"
            },
            {
                "id": 1,
                "second": "1247",
                "busId": "3264",
                "branch": "39",
                "tripTag": "45227729",
                "epochTime": "1672202531994"
            },
            {
                "id": 2,
                "second": "2447",
                "busId": "9234",
                "branch": "39",
                "tripTag": "45227728",
                "epochTime": "1672203731994"
            }
        ],
        "stopTag": 12925
    }
]
</p>
`;

exports[`eta parsing 1 predictions 2 directions 1`] = `
<p>
  [
    {
        "line": "39",
        "stopName": "Finch Station",
        "routeName": "",
        "etas": [
            {
                "id": 0,
                "second": "611",
                "busId": "3124",
                "branch": "39A",
                "tripTag": "45227587",
                "epochTime": "1672202160000"
            },
            {
                "id": 1,
                "second": "1811",
                "busId": "3281",
                "branch": "39A",
                "tripTag": "45227586",
                "epochTime": "1672203360000"
            },
            {
                "id": 2,
                "second": "3011",
                "busId": "3234",
                "branch": "39A",
                "tripTag": "45227585",
                "epochTime": "1672204560000"
            }
        ],
        "stopTag": 14211
    },
    {
        "line": "39",
        "stopName": "Finch Station",
        "routeName": "",
        "etas": [
            {
                "id": 0,
                "second": "11",
                "busId": "9227",
                "branch": "39B",
                "tripTag": "45227579",
                "epochTime": "1672201560000"
            },
            {
                "id": 1,
                "second": "1211",
                "busId": "3283",
                "branch": "39B",
                "tripTag": "45227578",
                "epochTime": "1672202760000"
            },
            {
                "id": 2,
                "second": "2411",
                "busId": "3109",
                "branch": "39B",
                "tripTag": "45227577",
                "epochTime": "1672203960000"
            }
        ],
        "stopTag": 14211
    }
]
</p>
`;

exports[`eta parsing 2 predictions 2 direction 1`] = `
<p>
  [
    {
        "line": "39",
        "stopName": "Finch Ave East At Wilfred Ave",
        "routeName": "Finch East towards Neilson",
        "etas": [
            {
                "id": 0,
                "second": "338",
                "busId": "3124",
                "branch": "39A",
                "tripTag": "45227587",
                "epochTime": "1672202317976"
            },
            {
                "id": 1,
                "second": "1538",
                "busId": "3281",
                "branch": "39A",
                "tripTag": "45227586",
                "epochTime": "1672203517976"
            },
            {
                "id": 2,
                "second": "2738",
                "busId": "3234",
                "branch": "39A",
                "tripTag": "45227585",
                "epochTime": "1672204718036"
            }
        ],
        "stopTag": 8607
    },
    {
        "line": "39",
        "stopName": "Finch Ave East At Wilfred Ave",
        "routeName": "Finch East towards Old Finch and Morningview",
        "etas": [
            {
                "id": 0,
                "second": "447",
                "busId": "9227",
                "branch": "39B",
                "tripTag": "45227579",
                "epochTime": "1672202426996"
            },
            {
                "id": 1,
                "second": "1289",
                "busId": "3283",
                "branch": "39B",
                "tripTag": "45227578",
                "epochTime": "1672203268756"
            },
            {
                "id": 2,
                "second": "2489",
                "busId": "3109",
                "branch": "39B",
                "tripTag": "45227577",
                "epochTime": "1672204468756"
            }
        ],
        "stopTag": 8607
    }
]
</p>
`;

exports[`eta parsing empty result 1`] = `
<p>
  []
</p>
`;

exports[`eta parsing no ETA 1`] = `
<p>
  [
    {
        "line": "",
        "stopName": "480 Gordon Baker Rd",
        "routeName": "",
        "etas": [],
        "stopTag": 7900
    }
]
</p>
`;
